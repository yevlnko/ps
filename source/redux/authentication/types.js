export const types = Object.freeze({
    // authentication
    AUTHENTICATE:         "AUTHENTICATE",
    AUTHENTICATE_SUCCESS: "AUTHENTICATE_SUCCESS",
    AUTHENTICATE_FAIL:    "AUTHENTICATE_FAIL",

    // SIGNUP
    SIGNUP:         "SIGNUP",
    SIGNUP_SUCCESS: "SIGNUP_SUCCESS",
    SIGNUP_FAIL:    "SIGNUP_FAIL",

    // logout
    LOGOUT:         "LOGOUT",
    LOGOUT_SUCCESS: "LOGOUT_SUCCESS",
    LOGOUT_FAIL:    "LOGOUT_FAIL",

    // login
    LOGIN:         "LOGIN",
    LOGIN_SUCCESS: "LOGIN_SUCCESS",
    LOGIN_FAIL:    "LOGIN_FAIL",
});
