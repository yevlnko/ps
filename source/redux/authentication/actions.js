// Instruments
import { types } from "./types";

export const authActions = Object.freeze({
    login: (credentials) => ({
        type:    types.LOGIN,
        payload: credentials,
    }),
    loginSuccess: () => ({
        type: types.LOGIN_SUCCESS,
    }),
    loginFail: (error) => ({
        type:    types.LOGIN_FAIL,
        payload: error,
        error:   true,
    }),

    authenticate: (token) => ({
        type:    types.AUTHENTICATE,
        payload: token,
    }),
    authenticateSuccess: () => ({
        type: types.AUTHENTICATE_SUCCESS,
    }),
    authenticateFail: (error) => ({
        type:    types.AUTHENTICATE_FAIL,
        payload: error,
        error:   true,
    }),

    signup: (credentials) => ({
        type:    types.SIGNUP,
        payload: credentials,
    }),
    signupSuccess: () => ({
        type: types.SIGNUP_SUCCESS,
    }),
    signupFail: (error) => ({
        type:    types.SIGNUP_FAIL,
        payload: error,
        error:   true,
    }),

    logout: () => ({
        type: types.LOGOUT,
    }),
    logoutSuccess: () => ({
        type: types.LOGOUT_SUCCESS,
    }),
    logoutFail: (error) => ({
        type:    types.LOGOUT_FAIL,
        payload: error,
        error:   true,
    }),
});
