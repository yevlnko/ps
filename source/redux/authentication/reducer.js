// Core
import { Map } from 'immutable';

// Instruments
import { types } from "./types";

const initialState = Map({
    isAuthenticated: false,
});

export const authReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.LOGIN_SUCCESS:
        case types.SIGNUP_SUCCESS:
        case types.AUTHENTICATE_SUCCESS:
            return state.set('isAuthenticated', true);

        case types.LOGOUT_SUCCESS:
            return state.set('isAuthenticated', false);


        default:
            return state;
    }
};
