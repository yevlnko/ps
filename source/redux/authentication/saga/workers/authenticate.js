// Core
import { call, put } from "redux-saga/effects";
import { actions } from "react-redux-form";

// Instruments
import { api } from "config";
import { authActions } from "redux/authentication/actions";
import { uiActions } from "redux/ui/actions";
import { profileActions } from "redux/profile/actions";

export function* callAuthenticateWorker ({ payload: token }) {
    try {
        const response = yield call(fetch, `${api}/user/login`, {
            method:  "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({ token }),
        });

        const { data: profile, message } = yield call([
            response,
            response.json
        ]);

        if (response.status !== 200) {
            if (response.status === 401) {
                yield localStorage.removeItem("token");

                return null;
            }
            throw new Error(message);
        }

        yield put(authActions.authenticateSuccess());
        yield put(profileActions.fillProfile(profile));

        yield put(
            actions.change("forms.user.profile.firstName", profile.firstName)
        );
        yield put(
            actions.change("forms.user.profile.lastName", profile.lastName)
        );
    } catch (err) {
        yield put(authActions.signupFail(err));
    } finally {
        yield put(uiActions.initializeSuccess(false));
    }
}
