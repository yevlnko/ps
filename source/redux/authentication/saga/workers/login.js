// Core
import { call, put } from 'redux-saga/effects';
import { replace } from 'react-router-redux';
import { actions } from 'react-redux-form';


// Instruments
import { api } from 'config';
import { authActions } from "redux/authentication/actions";
import { uiActions } from "redux/ui/actions";
import { profileActions } from "redux/profile/actions";
import { book } from "navigation/book";


export function* callLoginWorker ({ payload: credentials }) {

    try {
        yield put(uiActions.setAuthFetchingState(true));

        const response = yield call(fetch, `${api}/user/login`, {
            method:  'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(credentials),
        });

        const { data: profile, message } = yield call([response, response.json]);

        if (response.status !== 200) {
            throw new Error(message);
        }

        if (credentials.remember) {
            yield localStorage.setItem('token', profile.token);
        }

        yield put(authActions.loginSuccess());
        yield put(profileActions.fillProfile(profile));
        yield put(replace(book.feed));
        yield put(actions.reset('forms.login'));

    } catch (err) {

        yield put(authActions.loginFail(err));

    } finally {

        yield put(uiActions.setAuthFetchingState(false));

    }

}
