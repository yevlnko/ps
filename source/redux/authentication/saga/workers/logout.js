// Core
import { call, put } from "redux-saga/effects";
import { replace } from "react-router-redux";
import { actions } from "react-redux-form";

// Instruments
import { api } from "config";
import { authActions } from "redux/authentication/actions";
import { uiActions } from "redux/ui/actions";
import { profileActions } from "redux/profile/actions";
import { book } from "navigation/book";

export function* callLogoutWorker () {
    try {
        yield put(uiActions.setPostFetchingState(true));

        const token = yield localStorage.getItem("token");

        const response = yield call(fetch, `${api}/user/logout`, {
            method:  "GET",
            headers: {
                Authorization: token,
            },
        });

        if (response.status !== 204) {
            const { message } = yield call([response, response.json]);

            throw new Error(message);
        }

        yield put(authActions.logoutSuccess());
    } catch (err) {
        yield put(authActions.logoutFail(err));
    } finally {
        yield localStorage.removeItem("token");
        yield put(replace(book.feed));
        yield put(actions.reset("forms.login"));
        yield put(actions.reset("forms.profile"));
        yield put(profileActions.clearProfile());
        yield put(uiActions.setAuthFetchingState(false));
    }
}
