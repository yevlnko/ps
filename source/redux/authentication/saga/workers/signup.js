// Core
import { call, put } from 'redux-saga/effects';
import { replace } from 'react-router-redux';
import { actions } from 'react-redux-form';


// Instruments
import { api, groupID } from 'config';
import { authActions } from "redux/authentication/actions";
import { uiActions } from "redux/ui/actions";
import { profileActions } from "redux/profile/actions";
import { book } from "navigation/book";


export function* callSignUpWorker ({ payload: credentials }) {

    try {
        yield put(uiActions.setPostFetchingState(true));

        const response = yield call(fetch, `${api}/user/${groupID}`, {
            method:  'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(credentials),
        });

        const { data: profile, message } = yield call([response, response.json]);

        if (response.status !== 200) {
            throw new Error(message);
        }

        yield localStorage.setItem('token', profile.token);

        yield put(authActions.signupSuccess());
        yield put(profileActions.fillProfile(profile));
        yield put(replace(book.feed));
        yield put(actions.reset('forms.signup'));

    } catch (err) {

        yield put(authActions.signupFail(err));

    } finally {

        yield put(uiActions.setAuthFetchingState(false));

    }

}
