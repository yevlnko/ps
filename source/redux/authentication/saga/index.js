import { takeEvery } from 'redux-saga/effects';

// Instruments
import { types } from "redux/authentication/types";
import { callSignUpWorker } from "./workers/signup";
import { callLoginWorker } from "./workers/login";
import { callAuthenticateWorker } from "./workers/authenticate";
import { callLogoutWorker } from "./workers/logout";


export const authWatchers = Object.freeze({
    * watchSignUp () {
        yield takeEvery(types.SIGNUP, callSignUpWorker);
    },
    * watchLogin () {
        yield takeEvery(types.LOGIN, callLoginWorker);
    },
    * watchAuthenticate () {
        yield takeEvery(types.AUTHENTICATE, callAuthenticateWorker);
    },
    * watchLogout () {
        yield takeEvery(types.LOGOUT, callLogoutWorker);
    },
});
