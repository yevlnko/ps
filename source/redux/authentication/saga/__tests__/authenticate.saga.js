// Core
import { cloneableGenerator } from 'redux-saga/utils';

// Instruments
import { authActions } from "redux/authentication/actions";
import { callAuthenticateWorker } from "../workers/authenticate";

const authenticateActions = authActions.authenticate(__.token);
const saga = cloneableGenerator(callAuthenticateWorker)(authenticateActions);

describe('authenticate saga:', () => {

    test('should yield fetch', () => {
        expect(
            saga.next().value
        ).toMatchSnapshot();
    });

    test('should handle 401 status', () => {
        const clone = saga.clone();

        localStorage.setItem('token', __.token);

        // should yield transform data
        expect(clone.next(__.fetchResponseFail).value).toMatchSnapshot();

        // should remove token from localStorage
        expect(clone.next(__.responseDataFail).value).toBeUndefined();
        expect(localStorage.getItem('token')).toBeNull();

        // should call INITIALIZE_SUCCESS
        expect(clone.next().value).toMatchSnapshot();

        // should finish with null
        expect(clone.next().value).toBeNull();
        expect(clone.next().done).toBe(true);
    });

    test('should handle !== 200 response status', () => {
        const clone = saga.clone();

        const responseStatus400 = { ...__.fetchResponseFail, status: 400 };

        // should yield transformed data
        expect(clone.next(responseStatus400).value).toMatchSnapshot();

        // should put authenticateFail
        expect(clone.next(__.responseDataFail).value).toMatchSnapshot();

        // should put initializeSuccess
        expect(clone.next().value).toMatchSnapshot();

        // should finish
        expect(clone.next().done).toBe(true);
    });


    test('should handle 200 response status', () => {
        // should yield transformed data
        expect(saga.next(__.fetchResponseSuccess).value).toMatchSnapshot();

        // should yield authenticateSuccess
        expect(saga.next(__.responseDataSuccess).value).toMatchSnapshot();

        // should yield fillProfile
        expect(saga.next().value).toMatchSnapshot();

        // should yield redux-form
        expect(saga.next().value).toMatchSnapshot();

        // should yield redux-form
        expect(saga.next().value).toMatchSnapshot();

        // should yield initializeSuccess
        expect(saga.next().value).toMatchSnapshot();


        // should finish
        expect(saga.next().done).toBe(true);
    });

});
