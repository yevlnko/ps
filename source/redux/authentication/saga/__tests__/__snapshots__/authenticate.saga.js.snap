// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate saga: should handle !== 200 response status 1`] = `
Object {
  "CALL": Object {
    "args": Array [],
    "context": Object {
      "json": [MockFunction],
      "status": 400,
    },
    "fn": [MockFunction],
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle !== 200 response status 2`] = `
Object {
  "PUT": Object {
    "action": Object {
      "error": true,
      "payload": [Error: TEST_ERROR_MESSAGE.],
      "type": "SIGNUP_FAIL",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle !== 200 response status 3`] = `
Object {
  "PUT": Object {
    "action": Object {
      "type": "INITIALIZE_SUCCESS",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 1`] = `
Object {
  "CALL": Object {
    "args": Array [],
    "context": Object {
      "json": [MockFunction],
      "status": 200,
    },
    "fn": [MockFunction],
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 2`] = `
Object {
  "PUT": Object {
    "action": Object {
      "type": "AUTHENTICATE_SUCCESS",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 3`] = `
Object {
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "avatar": "TEST_AVATAR",
        "firstName": "Walter",
        "id": "TEST_ID",
        "lastName": "White",
        "token": "TEST_TOKEN",
      },
      "type": "FILL_PROFILE",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 4`] = `
Object {
  "PUT": Object {
    "action": Object {
      "external": true,
      "model": "forms.user.profile.firstName",
      "multi": false,
      "silent": false,
      "type": "rrf/change",
      "value": "Walter",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 5`] = `
Object {
  "PUT": Object {
    "action": Object {
      "external": true,
      "model": "forms.user.profile.lastName",
      "multi": false,
      "silent": false,
      "type": "rrf/change",
      "value": "White",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 200 response status 6`] = `
Object {
  "PUT": Object {
    "action": Object {
      "type": "INITIALIZE_SUCCESS",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 401 status 1`] = `
Object {
  "CALL": Object {
    "args": Array [],
    "context": Object {
      "json": [MockFunction],
      "status": 401,
    },
    "fn": [MockFunction],
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should handle 401 status 2`] = `
Object {
  "PUT": Object {
    "action": Object {
      "type": "INITIALIZE_SUCCESS",
    },
    "channel": null,
  },
  Symbol(@@redux-saga/IO): true,
}
`;

exports[`authenticate saga: should yield fetch 1`] = `
Object {
  "CALL": Object {
    "args": Array [
      "https://lab.lectrum.io/redux/api/user/login",
      Object {
        "body": "{\\"token\\":\\"TEST_TOKEN\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [MockFunction],
  },
  Symbol(@@redux-saga/IO): true,
}
`;
